### PR TITLE
Feat: support union JavaScript types for bindings

### PR DESCRIPTION
**Pull request by andycall**
1. [x] Support generate union types from typings [For core devs].

If you write `fillStyle: string | CanvasGradient | null;` at typings, the code generator will auto generate the union C++ class for you.

```c++
class QJSUnionDomStringCanvasGradient final : public UnionBase {
 public:
  using ImplType = std::shared_ptr<QJSUnionDomStringCanvasGradient>;
 ....
};
```

Then you can read the dynamic values from this union class in your C++  code without concerned of the JavaScript types.

```c++
void CanvasRenderingContext2D::setFillStyle(const std::shared_ptr<QJSUnionDomStringCanvasGradient>& style, ExceptionState& exception_state) {
  NativeValue value = Native_NewNull();

  if (style->IsDomString()) {
    value = NativeValueConverter<NativeTypeString>::ToNativeValue(style->GetAsDomString());
  } else if (style->IsCanvasGradient()) {
    value = NativeValueConverter<NativeTypePointer<CanvasGradient>>::ToNativeValue(style->GetAsCanvasGradient());
  }

  //....
}
```

### CODE REVIEW

**Review by devjiangzhou — APPROVED** [review]

File: bridge/scripts/code_generator/bin/code_generator.js [hunk 1]
```diff
@@ -72,6 +74,14 @@ function genCodeFromTypeDefine() {
     wirteFileIfChanged(genFilePath + '.h', result.header);
     wirteFileIfChanged(genFilePath + '.cc', result.source);
   }
+
+  let unionTypes = Array.from(unionTypeCollector.types);
```
Comment (XGHeaven) [inline bridge/scripts/code_generator/bin/code_generator.js:78]: - unionTypes 的内的每个元素自身需要根据类型名称 sort 一下，避免 `string | double[]` 与 `double[] | string` 创建两个  class
- collector types 的 Set 类型是不能去重对象的，所以还需要按照文件名进行一次去重，否则会对同一个文件写两次，虽然结果没啥影响，就是性能会差一些。

File: bridge/test/webf_test_env.cc [hunk 2]
```diff
@@ -188,6 +188,10 @@ void TEST_flushUICommand(int32_t contextId) {
   clearUICommandItems(reinterpret_cast<void*>(page));
 }
 
+void TEST_CreateBindingObject(int32_t context_id, void* native_binding_object, int32_t type, void* args, int32_t argc) {
+
```
Comment (XGHeaven) [inline bridge/test/webf_test_env.cc:192]: remove unused test

File: webf/lib/src/bridge/native_value.dart [hunk 3]
```diff
@@ -72,65 +82,41 @@ dynamic fromNativeValue(Pointer<NativeValue> nativeValue) {
...
-typedef ToNativeValueCallback = void Function(Pointer<NativeValue> target,
-    // ignore: avoid_annotating_with_dynamic
-    {dynamic value, BindingObject? ownerBindingObject});
-
-final HashMap<Type, ToNativeValueCallback> _nativeToValueMap = HashMap();
-bool _nativeToNativeValueMapInit = false;
-
-void _initNativeToValueMap() {
-  _nativeToValueMap[null.runtimeType] = (target, {value, ownerBindingObject}) {
+void toNativeValue(Pointer<NativeValue> target, value, [BindingObject? ownerBindingObject]) {
```
Comment (XGHeaven) [inline webf/lib/src/bridge/native_value.dart:85]: 这块的实现为啥又 revert 回去了？是不是本地分支乱了？

File: webf/lib/src/geometry/dom_matrix.dart [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/*
+ * Copyright (C) 2022-present The WebF authors. All rights reserved.
+ */
+
+import 'package:webf/foundation.dart';
+import 'dom_matrix_readonly.dart';
+
+class DOMMatrix extends DOMMatrixReadonly {
+  DOMMatrix(BindingContext context, List<dynamic> domMatrixInit): super(context, domMatrixInit) {
+    print('domMatrix init: $domMatrixInit');
```
Comment (XGHeaven) [inline webf/lib/src/geometry/dom_matrix.dart:10]: delete this log?

File: bridge/CMakeLists.txt [hunk 5]
```diff
@@ -18,7 +18,7 @@ else ()
 endif ()
 
 execute_process(
-  COMMAND bash "-c" "npm install"
+  COMMAND bash "-c" "if [ ! -d \"node_modules\" ]; then npm install; fi"
```
Comment (XGHeaven) [inline bridge/CMakeLists.txt:21]: 这一行也要 revert 回去吧

File: webf/lib/src/bridge/native_value.dart (hunk 3 above)
Comment (andycall) [reply webf/lib/src/bridge/native_value.dart:85]: 发现性能不太好

File: bridge/scripts/code_generator/bin/code_generator.js (hunk 1 above)
Comment (andycall) [reply bridge/scripts/code_generator/bin/code_generator.js:78]: OK

File: bridge/test/webf_test_env.cc (hunk 2 above)
Comment (andycall) [reply bridge/test/webf_test_env.cc:192]: This function are simulate for dart methods.

File: webf/lib/src/geometry/dom_matrix.dart (hunk 4 above)
Comment (andycall) [reply webf/lib/src/geometry/dom_matrix.dart:10]: It's fine.

**Review by XGHeaven — APPROVED** [review]

**Pull request merged**